File: internal/frontier/frontier.go
package frontier

import (
	"fmt"
	"sort"

	"github.com/JarenKempton/nav/internal/domain"
)

type Scope struct {
	Workspace *domain.WorkspaceRef
	Group     *domain.GroupRef
	Map       *domain.MapRef
	Ticket    *domain.TicketRef
}

type Options struct {
	AvailableStatuses  map[string]bool
	IncludeStaleClaims bool
}

func Evaluate(tickets []domain.Ticket, scope Scope, options Options) ([]domain.Ticket, error) {
	byRef := make(map[domain.TicketRef]domain.Ticket, len(tickets))
	for _, ticket := range tickets {
		byRef[ticket.Ref] = ticket
	}
	eligible := make([]domain.Ticket, 0)
	for _, ticket := range tickets {
		if !inScope(ticket, scope) || ticket.State != domain.TicketOpen || !options.AvailableStatuses[ticket.Status] || ticket.Assignee != nil {
			continue
		}
		blocked := false
		for _, dependency := range ticket.Dependencies {
			if dependency.Blocked != ticket.Ref {
				continue
			}
			blocker, ok := byRef[dependency.Blocking]
			if !ok {
				return nil, fmt.Errorf("ticket %s references unknown blocker %s", ticket.Ref, dependency.Blocking)
			}
			if blocker.State != domain.TicketClosed {
				blocked = true
				break
			}
		}
		if !blocked {
			eligible = append(eligible, ticket)
		}
	}
	sort.SliceStable(eligible, func(i, j int) bool {
		if eligible[i].Order == eligible[j].Order {
			return eligible[i].Ref < eligible[j].Ref
		}
		return eligible[i].Order < eligible[j].Order
	})
	return eligible, nil
}

func inScope(ticket domain.Ticket, scope Scope) bool {
	if scope.Ticket != nil {
		return ticket.Ref == *scope.Ticket
	}
	if scope.Map != nil {
		return ticket.Map == *scope.Map
	}
	if scope.Group != nil {
		return ticket.Group != nil && *ticket.Group == *scope.Group
	}
	return true
}

func Select(tickets []domain.Ticket, policy string) (domain.Ticket, error) {
	if len(tickets) == 0 {
		return domain.Ticket{}, fmt.Errorf("frontier is empty")
	}
	switch policy {
	case "first":
		return tickets[0], nil
	case "highest-priority":
		selected := tickets[0]
		for _, ticket := range tickets[1:] {
			if ticket.Priority > selected.Priority {
				selected = ticket
			}
		}
		return selected, nil
	default:
		return domain.Ticket{}, fmt.Errorf("unknown or missing noninteractive selection policy %q", policy)
	}
}

File: src/frontier.ts
import type { GroupRef, MapRef, Ticket, TicketRef, WorkspaceRef } from "./domain.ts";
import { parseRef, workspaceRefOf } from "./reference.ts";

export interface FrontierScope {
  workspace?: WorkspaceRef;
  group?: GroupRef;
  map?: MapRef;
  ticket?: TicketRef;
}

export interface FrontierOptions {
  availableStatuses: ReadonlySet<string>;
}

export function evaluateFrontier(
  tickets: readonly Ticket[],
  scope: FrontierScope,
  options: FrontierOptions,
): Ticket[] {
  const normalized = normalizeTrackerTickets(tickets);
  const normalizedScope = normalizeScope(scope);
  validateScope(normalizedScope, normalized.workspace);
  const byRef = new Map<TicketRef, Ticket>(
    normalized.tickets.map((ticket) => [ticket.ref, ticket]),
  );
  const eligible: Ticket[] = [];

  for (const ticket of normalized.tickets) {
    if (!isInScope(ticket, normalizedScope)) continue;
    if (ticket.state !== "open") continue;
    if (!options.availableStatuses.has(ticket.status)) continue;
    if (ticket.assignee !== undefined) continue;

    let blocked = false;
    for (const dependency of ticket.dependencies ?? []) {
      const blocker = byRef.get(dependency.blocking);
      if (!blocker) {
        throw new Error(`Ticket ${ticket.ref} references unknown blocker ${dependency.blocking}`);
      }
      if (blocker.state !== "closed") {
        blocked = true;
        break;
      }
    }
    if (!blocked) eligible.push(ticket);
  }

  // Modern Array#sort is stable. Equal order values therefore retain the tracker
  // adapter's input order instead of inventing a reference-based ordering.
  return eligible.toSorted((left, right) => left.order - right.order);
}

export interface NormalizedTrackerTickets {
  workspace?: WorkspaceRef;
  tickets: Ticket[];
}

/**
 * Validates the normalized tracker boundary before core evaluates policy.
 * V1 intentionally fails closed on partial or cross-workspace dependency graphs.
 */
export function normalizeTrackerTickets(tickets: readonly Ticket[]): NormalizedTrackerTickets {
  const seen = new Set<TicketRef>();
  let workspace: WorkspaceRef | undefined;
  const normalized: Ticket[] = [];

  for (const ticket of tickets) {
    if (!ticket || typeof ticket !== "object") throw new Error("Invalid normalized ticket");
    requireKind(ticket.ref, "ticket");
    requireKind(ticket.map, "map");
    if (ticket.group !== undefined) requireKind(ticket.group, "group");
    if (!["task", "research", "prototype", "decision"].includes(ticket.kind)) {
      throw new Error(`Ticket ${ticket.ref} has an unsupported kind: ${ticket.kind}`);
    }
    if (ticket.state !== "open" && ticket.state !== "closed") {
      throw new Error(`Ticket ${ticket.ref} has an unsupported state: ${ticket.state}`);
    }
    if (typeof ticket.status !== "string" || ticket.status.length === 0) {
      throw new Error(`Ticket ${ticket.ref} has an invalid status`);
    }
    if (!Number.isFinite(ticket.order)) {
      throw new Error(`Ticket ${ticket.ref} has a non-finite tracker order`);
    }
    if (seen.has(ticket.ref)) throw new Error(`Duplicate ticket reference: ${ticket.ref}`);
    seen.add(ticket.ref);

    const ticketWorkspace = workspaceRefOf(ticket.ref) as WorkspaceRef;
    if (workspace === undefined) workspace = ticketWorkspace;
    if (workspace !== ticketWorkspace) {
      throw new Error(`Cross-workspace frontier input is not supported: ${ticket.ref}`);
    }
    requireSameWorkspace(ticket.ref, ticket.map);
    if (ticket.group !== undefined) requireSameWorkspace(ticket.ref, ticket.group);

    for (const dependency of ticket.dependencies ?? []) {
      requireKind(dependency.blocking, "ticket");
      requireKind(dependency.blocked, "ticket");
      if (dependency.kind !== "blocks") {
        throw new Error(`Ticket ${ticket.ref} has an unsupported dependency kind`);
      }
      if (dependency.blocked !== ticket.ref) {
        throw new Error(
          `Ticket ${ticket.ref} contains a dependency owned by ${dependency.blocked}`,
        );
      }
      requireSameWorkspace(ticket.ref, dependency.blocking);
    }
    normalized.push({
      ...ticket,
      ...(ticket.dependencies
        ? { dependencies: ticket.dependencies.map((item) => ({ ...item })) }
        : {}),
    });
  }

  for (const ticket of normalized) {
    for (const dependency of ticket.dependencies ?? []) {
      if (!seen.has(dependency.blocking)) {
        throw new Error(`Ticket ${ticket.ref} references unknown blocker ${dependency.blocking}`);
      }
    }
  }

  return { ...(workspace ? { workspace } : {}), tickets: normalized };
}

export function selectFrontierTicket(tickets: readonly Ticket[], policy: string): Ticket {
  const first = tickets[0];
  if (!first) throw new Error("Frontier is empty");
  if (policy === "first") return first;
  if (policy === "highest-priority") {
    return tickets.reduce((selected, ticket) =>
      (ticket.priority ?? 0) > (selected.priority ?? 0) ? ticket : selected,
    );
  }
  throw new Error(`Unknown or missing noninteractive selection policy: ${JSON.stringify(policy)}`);
}

function isInScope(ticket: Ticket, scope: FrontierScope): boolean {
  if (scope.ticket !== undefined) return ticket.ref === scope.ticket;
  if (scope.map !== undefined) return ticket.map === scope.map;
  if (scope.group !== undefined) return ticket.group === scope.group;
  if (scope.workspace !== undefined) return workspaceRefOf(ticket.ref) === scope.workspace;
  return true;
}

function requireKind(reference: string, expected: "workspace" | "group" | "map" | "ticket"): void {
  const actual = parseRef(reference).kind;
  if (actual !== expected) {
    throw new Error(`Expected ${expected} reference, received ${actual}: ${reference}`);
  }
}

function requireSameWorkspace(owner: string, related: string): void {
  if (workspaceRefOf(owner) !== workspaceRefOf(related)) {
    throw new Error(`Cross-workspace relationship is not supported: ${owner} -> ${related}`);
  }
}

function validateScope(scope: FrontierScope, workspace: WorkspaceRef | undefined): void {
  const references = [scope.workspace, scope.group, scope.map, scope.ticket].filter(
    (item): item is WorkspaceRef | GroupRef | MapRef | TicketRef => item !== undefined,
  );
  if (references.length > 1) throw new Error("Frontier scope must contain at most one reference");
  const reference = references[0];
  if (
    reference !== undefined &&
    workspace !== undefined &&
    workspaceRefOf(reference) !== workspace
  ) {
    throw new Error(`Frontier scope ${reference} is outside workspace ${workspace}`);
  }
}

function normalizeScope(scope: FrontierScope): FrontierScope {
  return {
    ...(scope.workspace === undefined
      ? {}
      : { workspace: normalizeScopeRef(scope.workspace, "workspace") as WorkspaceRef }),
    ...(scope.group === undefined
      ? {}
      : { group: normalizeScopeRef(scope.group, "group") as GroupRef }),
    ...(scope.map === undefined ? {} : { map: normalizeScopeRef(scope.map, "map") as MapRef }),
    ...(scope.ticket === undefined
      ? {}
      : { ticket: normalizeScopeRef(scope.ticket, "ticket") as TicketRef }),
  };
}

function normalizeScopeRef(
  reference: string,
  expected: "workspace" | "group" | "map" | "ticket",
): string {
  const parsed = parseRef(reference);
  if (parsed.kind !== expected) {
    throw new Error(`Expected ${expected} scope reference, received ${parsed.kind}: ${parsed.raw}`);
  }
  return parsed.raw;
}
